pkg/manager/member/master_scaler.go

```diff
@@ -58,31 +103,97 @@ func (s masterScaler) ScaleIn(meta metav1.Object, actual *apps.StatefulSet, desi
 
 	ns := tc.GetNamespace()
 	tcName := tc.GetName()
+	stsName := actual.GetName()
+
 	if !tc.Status.Master.Synced {
-		return errors.Errorf("tiflow cluster: %s/%s's tiflow-master status sync failed, can't scale out now", ns, tcName)
+		return errors.Errorf("tiflow cluster: %s/%s's tiflow-master status sync failed, can't scale out now",
+			ns, tcName)
 	}
 
-	ordinal := *actual.Spec.Replicas - 1
-	memberName := ordinalPodName(v1alpha1.TiFlowMasterMemberType, tcName, ordinal)
+	klog.Infof("start to scaling down tiflow-master statefulSet %s for [%s/%s]",
+		stsName, ns, tcName)
 
-	if !tc.Status.Master.Synced {
-		return fmt.Errorf("tiflow cluster: %s/%s's tiflow-master status sync failed, can't scale in now", ns, tcName)
+	down := *actual.Spec.Replicas - *desired.Spec.Replicas
+	current := *actual.Spec.Replicas
+	ctx := context.TODO()
+
+	for i := down; i > 0; i-- {
+		klog.Infof("scaling down statefulSet  %s of master, current: %d, desired: %d",
+			stsName, current, current-1)
+
+		if err := s.EvictLeader(tc, current-1); err != nil {
+			return nil
+		}
+
+		if err := s.SetReplicas(ctx, actual, uint(current-1)); err != nil {
+			return err
+		}
+
+		if err := s.WaitUntilHealthy(ctx, uint(current-1)); err != nil {
```

Comment:
logic need to be changed

Reply:
I think, If we have migrated the leader, we should make sure that the current leader is the correct one.